pulser/simulation/simresults.py

```diff
@@ -74,15 +72,14 @@ def plot(self, op: qutip.Qobj, fmt: str = '.', label: str = '') -> None:
         pass
 
     @abstractmethod
-    def _calc_weights(self, t: int) -> ArrayLike:
+    def _calc_weights(self, t: float) -> ArrayLike:
         """Computes the bitstring probabilities for sampled states."""
         pass
 
     def expect(self, obs_list: Sequence[Union[qutip.Qobj, ArrayLike]]
                ) -> list[Union[float, complex, ArrayLike]]:
         """Returns the expectation values of operators in obs_list."""
-        states = [self.get_state(t) for
-                  t in range(len(self._results))]
+        states = [self.get_state(t) for t in self.sim_times]
```

Comment:
I understand this is useful for modularity, but it seems terribly inefficient... A nice solution might be to do move this to the `states` property as we discussed: in that way, this loop only happens for `NoisyResults` and here you would just get `states` from `self.states`.